File: src/lexflow/api/routers/telemetry.py
"""``/api/v1/telemetry`` — opt-in usage telemetry (#74).

Two endpoints:

* ``GET  /telemetry/status`` — does the backend currently accept events?
* ``POST /telemetry/events`` — submit a batch (max 50) of telemetry
  records. Accepted regardless of the backend gate; persisted only
  when ``LEXFLOW_TELEMETRY_ENABLED=1``.

The SPA gates client-side too (a Zustand flag tied to a Settings
toggle), so a request only reaches this router if both the user and
the operator have opted in.
"""

from __future__ import annotations

import logging

from fastapi import APIRouter, status
from pydantic import BaseModel, Field

from lexflow.core.telemetry import (
    TelemetryEvent,
    TelemetryStatus,
    is_enabled,
    record_events,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/telemetry", tags=["Telemetry"])

_MAX_EVENTS_PER_BATCH = 50


class TelemetryBatch(BaseModel):
    """Inbound batch shape."""

    events: list[TelemetryEvent] = Field(default_factory=list, max_length=_MAX_EVENTS_PER_BATCH)


class TelemetryIngestResponse(BaseModel):
    """Per-request ack — how many of the batch were actually persisted."""

    accepted: int
    enabled: bool


@router.get(
    "/status",
    response_model=TelemetryStatus,
    summary="Does the backend currently persist telemetry events?",
)
def get_status() -> TelemetryStatus:
    """Return the live backend opt-in state.

    The SPA reads this on boot to decide whether to show the privacy
    toggle as available — if the env knob is unset the operator hasn't
    enabled it server-side, so even an opt-in user wouldn't generate
    any events.
    """
    return TelemetryStatus(enabled=is_enabled())


@router.post(
    "/events",
    response_model=TelemetryIngestResponse,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Submit a batch of telemetry events (silently dropped when disabled).",
)
def ingest_events(batch: TelemetryBatch) -> TelemetryIngestResponse:
    """Accept a batch; persist only when the backend gate is on.

    Returns 202 either way — the frontend doesn't need to branch on the
    server's gate, and we don't want telemetry firing to surface as an
    error in the UI's request log.
    """
    accepted = record_events(batch.events)
    return TelemetryIngestResponse(accepted=accepted, enabled=is_enabled())

File: frontend/src/lib/api/telemetry.ts
/**
 * `liveApi.telemetry` — backend opt-in status + event submission (#331).
 *
 * The product has two independent gates:
 *
 * 1. **Backend** (`LEXFLOW_TELEMETRY_ENABLED=1`): if the operator
 *    hasn't enabled it, the server returns 202 + `accepted: 0` and
 *    nothing reaches disk. The SPA reads `status()` to show the
 *    user whether their consent is even meaningful right now.
 * 2. **User** (Zustand `telemetryConsent`): if the user hasn't
 *    opted in, the SPA doesn't POST events at all.
 *
 * Events only flow when BOTH are on.
 */

import { http } from './http';

export interface TelemetryStatus {
  enabled: boolean;
}

export interface TelemetryEvent {
  name: string;
  props?: Record<string, unknown>;
}

export interface TelemetryIngestResponse {
  accepted: number;
  enabled: boolean;
}

export const liveTelemetryApi = {
  /** Read the backend opt-in gate. Cheap; safe to poll on Settings open. */
  status: async (): Promise<TelemetryStatus> => {
    return http<TelemetryStatus>('/telemetry/status');
  },
  /**
   * Submit a batch. The backend returns 202 either way; ``accepted``
   * reports how many actually reached disk. The SPA fires this only
   * after the user has opted in (Zustand ``telemetryConsent``).
   */
  events: async (events: TelemetryEvent[]): Promise<TelemetryIngestResponse> => {
    return http<TelemetryIngestResponse>('/telemetry/events', {
      method: 'POST',
      body: JSON.stringify({ events }),
      headers: { 'Content-Type': 'application/json' },
    });
  },
};
